### PR TITLE
Use Marshmallow to load/save config files

### PR DESCRIPTION
**Pull request by agateau-gg**
## Description

This PR is the Marshmallow version of #240. I switched to Marshmallow after realizing ggshield already indirectly depends on it, via pygitguardian.

## What has been done

- Some preliminary work: splitting test files for configuration classes.
- Add Marshmallow schema classes for our configuration classes.

Note that this PR targets the iac branch, to avoid disrupting the release.

### CODE REVIEW

File: ggshield/core/config/auth_config.py [hunk 1]
```diff
@@ -137,3 +152,6 @@ def get_instance_token(self, instance_name: str) -> str:
         if instance.account is None:
             raise MissingTokenError(instance=instance_name)
         return instance.account.token
+
+
+AuthConfigSchema = marshmallow_dataclass.class_schema(AuthConfig)
```
Comment (zhangmatthieu) [inline ggshield/core/config/auth_config.py:157]: I think we could use the decorators `@dataclass` https://pypi.org/project/marshmallow-dataclass/ to simplify the code

Comment (zhangmatthieu) [reply ggshield/core/config/auth_config.py:157]: Ha, we already have a dataclass decorator, not sure how those two would interact with each other

File: ggshield/core/config/user_config.py [hunk 2]
```diff
@@ -32,23 +41,12 @@ class UserConfig(YAMLFileConfig):
...
-        try:
-            api_url = data.pop("api_url")
-        except KeyError:
-            pass
-        else:
-            if "instance" not in data:
-                data["instance"] = api_to_dashboard_url(api_url, warn=True)
-        return super(UserConfig, self).update_config(data)
-
     def save(self, config_path: str) -> None:
```
Comment (zhangmatthieu) [inline ggshield/core/config/user_config.py:44]: For my own understanding, are we writing on config files when calling save on the config?

Comment (agateau-gg) [reply ggshield/core/config/user_config.py:44]: Yes, that's what is happening.

File: ggshield/core/config/utils.py [hunk 3]
```diff
@@ -80,60 +65,37 @@ def get_global_path(filename: str) -> str:
...
-        p.parent.mkdir(parents=True, exist_ok=True)
-        with p.open("w") as f:
-            try:
-                stream = yaml.dump(data, indent=2, default_flow_style=False)
-                f.write(stream)
-            except Exception as e:
-                raise click.ClickException(
-                    f"Error while saving config in {path}:\n{str(e)}"
-                ) from e
+def update_from_other_instance(dst: Any, src: Any) -> None:
```
Comment (zhangmatthieu) [inline ggshield/core/config/utils.py:68]: nit: we don't have a test for this method, should we add one ?

**Review by zhangmatthieu — APPROVED** [review]

**Pull request merged**